`dartPluginClass` forces to export implementation of platform interface as public api

### Use case

Wasn't sure about type of Issue here, but decided to do "Feature Request" since I believe it can be new functionality based on source code from Flutter, which I'll provide later.

Question is about usage of `dartPluginClass` for providing dart-only implementations for Flutter plugins.

At the moment, when platform-specific implementation is provided with `dartPluginClass`, this implementation has to be exported as Public API. It is obligatory, as far as I understand, because it is not possible to set `fileName` (as for `web` implementation) with `dartPluginClass`. My observation based on [source-code of dart_plugin_registrant.dart template](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/flutter_plugins.dart#L621-L673) which doesn't allow other imports than `'package:{{pluginName}}/{{pluginName}}.dart'`

If implementation classes are not exported, build will brake on all platforms, except for web, because web has it's own mechanism.

Build error will look like this (this stack trace is from example repo for reproduction, I will provide link in the end.):

```
Launching lib/main.dart on macOS in debug mode...
--- xcodebuild: WARNING: Using the first of multiple matching destinations:
{ platform:macOS, arch:arm64, id:00006000-000849E41162801E, name:My Mac }
{ platform:macOS, arch:x86_64, id:00006000-000849E41162801E, name:My Mac }
.dart_tool/flutter_build/dart_plugin_registrant.dart:22:9: Error: Undefined name 'DartCounterImpl'.
        DartCounterImpl.registerWith();
        ^^^^^^^^^^^^^^^
.dart_tool/flutter_build/dart_plugin_registrant.dart:32:9: Error: Undefined name 'DartCounterImpl'.
        DartCounterImpl.registerWith();
        ^^^^^^^^^^^^^^^
.dart_tool/flutter_build/dart_plugin_registrant.dart:42:9: Error: Undefined name 'DartCounterImpl'.
        DartCounterImpl.registerWith();
        ^^^^^^^^^^^^^^^
.dart_tool/flutter_build/dart_plugin_registrant.dart:52:9: Error: Undefined name 'DartCounterImpl'.
        DartCounterImpl.registerWith();
        ^^^^^^^^^^^^^^^
.dart_tool/flutter_build/dart_plugin_registrant.dart:62:9: Error: Undefined name 'DartCounterImpl'.
        DartCounterImpl.registerWith();
        ^^^^^^^^^^^^^^^
Target kernel_snapshot_program failed: Exception

Command PhaseScriptExecution failed with a nonzero exit code
** BUILD FAILED **
```

**Steps to reproduce**:

1. Use sample project from here: https://github.com/Sameri11/dart_plugin_class_example_error
2. Take a look at `lib/dart_plugin_class_sample.dart`: there is big comment about issue.
3. Comment or delete [this line (26)](https://github.com/Sameri11/dart_plugin_class_example_error/blob/main/lib/dart_plugin_class_sample.dart#L26)
4. Go to `example`
5. Run `flutter run` for any platform other than `web`
6. See exception from above
7. Optionally: if your revert step 3 (deleting line) – build will work

**Real life example**

I discovered this behaviour contributing to `flutter_file_picker` [here](https://github.com/miguelpruivo/flutter_file_picker/pull/1567). Not asking to look at that PR, but it can provide a little more context, if someone wants to dive in. TL:DR – this PR could benefit if exporting implementations wasn't obligatory.

**Flutter doctor -v**

```
[✓] Flutter (Channel master, 3.24.0-1.0.pre.411, on macOS 14.3.1 23D60 darwin-arm64, locale en-RU)
    • Flutter version 3.24.0-1.0.pre.411 on channel master at /Users/samer/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a1a3b2ce3c (2 days ago), 2024-08-02 06:09:24 -0400
    • Engine revision b408111b97
    • Dart version 3.6.0 (build 3.6.0-108.0.dev)
    • DevTools version 2.38.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/samer/Library/Android/sdk
    • Platform android-35, build-tools 34.0.0
    • Java binary at: /Users/samer/Applications/Android Studio Koala
      2024.1.1.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Users/samer/Applications/Android Studio Koala 2024.1.1.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[!] Android Studio (version unknown)
    • Android Studio at /Users/samer/Applications/Android Studio Koala Feature Drop 2024.1.2 Canary
      5.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)

[✓] Android Studio (version 2024.1)
    • Android Studio at /Users/samer/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] IntelliJ IDEA Ultimate Edition (version EAP IU-242.20224.38)
    • IntelliJ at /Users/samer/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.91.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.94.0

[✓] Connected device (6 available)
    • sdk gphone64 arm64 (mobile)     • emulator-5554                        • android-arm64  •
      Android 15 (API 35) (emulator)
    • iPhone (Михаил) (mobile)        • 00008020-001254DA1E39002E            • ios            • iOS
      17.5.1 21F90
    • iPhone 15 (mobile)              • 9AD45704-B570-4B87-902F-233C3F7F2280 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-17-4 (simulator)
    • macOS (desktop)                 • macos                                • darwin-arm64   •
      macOS 14.3.1 23D60 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad                • darwin         •
      macOS 14.3.1 23D60 darwin-arm64
    • Chrome (web)                    • chrome                               • web-javascript •
      Google Chrome 126.0.6478.185

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

### Proposal

Provide some possibility not to export implementation classes when using `dartPluginClass`. I'm not expert here, but obvious option from my perspective is to respect `fileName` field in `pubspec.yaml`, as `web` already does.

--- a/packages/flutter_tools/lib/src/flutter_plugins.dart
+++ b/packages/flutter_tools/lib/src/flutter_plugins.dart
@@ -624,19 +624,19 @@ const String _dartPluginRegistryForNonWebTemplate = '''
 
 import 'dart:io'; // flutter_ignore: dart_io_import.
 {{#android}}
-import 'package:{{pluginName}}/{{pluginName}}.dart';
+import 'package:{{pluginName}}/{{dartFileName}}';
 {{/android}}
 {{#ios}}
-import 'package:{{pluginName}}/{{pluginName}}.dart';
+import 'package:{{pluginName}}/{{dartFileName}}';
 {{/ios}}
 {{#linux}}
-import 'package:{{pluginName}}/{{pluginName}}.dart';
+import 'package:{{pluginName}}/{{dartFileName}}';
 {{/linux}}
 {{#macos}}
-import 'package:{{pluginName}}/{{pluginName}}.dart';
+import 'package:{{pluginName}}/{{dartFileName}}';
 {{/macos}}
 {{#windows}}
-import 'package:{{pluginName}}/{{pluginName}}.dart';
+import 'package:{{pluginName}}/{{dartFileName}}';
 {{/windows}}
 
 @pragma('vm:entry-point')
@@ -1433,7 +1433,7 @@ bool _hasPluginInlineImpl(
 /// Determine if the plugin provides an inline Dart implementation.
 bool _hasPluginInlineDartImpl(Plugin plugin, String platformKey) {
   return plugin.pluginDartClassPlatforms[platformKey] != null &&
-      plugin.pluginDartClassPlatforms[platformKey] != 'none';
+      plugin.pluginDartClassPlatforms[platformKey]!.dartClass != 'none';
 }
 
 /// Get the resolved plugin [resolution] from the [candidates] serving as implementation for

--- a/packages/flutter_tools/lib/src/flutter_plugins.dart
+++ b/packages/flutter_tools/lib/src/flutter_plugins.dart
@@ -1432,8 +1432,8 @@ bool _hasPluginInlineImpl(
 
 /// Determine if the plugin provides an inline Dart implementation.
 bool _hasPluginInlineDartImpl(Plugin plugin, String platformKey) {
-  return plugin.pluginDartClassPlatforms[platformKey] != null &&
-      plugin.pluginDartClassPlatforms[platformKey]!.dartClass != 'none';
+  final DartPluginClassAndFilePair? platformInfo = plugin.pluginDartClassPlatforms[platformKey];
+  return platformInfo != null && platformInfo.dartClass != 'none';
 }
 
 /// Get the resolved plugin [resolution] from the [candidates] serving as implementation for

--- a/packages/flutter_tools/lib/src/platform_plugins.dart
+++ b/packages/flutter_tools/lib/src/platform_plugins.dart
@@ -13,6 +13,9 @@ const String kPluginClass = 'pluginClass';
 /// Constant for 'dartPluginClass' key in plugin maps.
 const String kDartPluginClass = 'dartPluginClass';
 
+/// Constant for 'dartPluginFile' key in plugin maps.
+const String kDartFileName = 'dartFileName';
+
 /// Constant for 'ffiPlugin' key in plugin maps.
 const String kFfiPlugin = 'ffiPlugin';
 
@@ -68,8 +71,8 @@ abstract class DarwinPlugin {
 /// - an implementation consisting of:
 ///   - the [package] and [pluginClass] that will be the entry point to the
 ///     plugin's native code, and/or
-///   - the [dartPluginClass] that will be the entry point for the plugin's
-///     Dart code
+///   - the [dartPluginClass] with optional [dartFileName] that will be
+///     the entry point for the plugin's Dart code
 /// is required.
 class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
   AndroidPlugin({
@@ -78,6 +81,7 @@ class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
     this.package,
     this.pluginClass,
     this.dartPluginClass,
+    this.dartFileName,
     bool? ffiPlugin,
     this.defaultPackage,
     required FileSystem fileSystem,
@@ -89,6 +93,7 @@ class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
         package = yaml['package'] as String?,
         pluginClass = yaml[kPluginClass] as String?,
         dartPluginClass = yaml[kDartPluginClass] as String?,
+        dartFileName = yaml[kDartFileName] as String?,
         ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
         defaultPackage = yaml[kDefaultPackage] as String?,
         _fileSystem = fileSystem;
@@ -125,6 +130,9 @@ class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
   /// The Dart plugin main class defined in pubspec.yaml, if any.
   final String? dartPluginClass;
 
+  /// Path to file in which dartPluginClass defined, if any.
+  final String? dartFileName;
+
   /// Is FFI plugin defined in pubspec.yaml.
   final bool ffiPlugin;
 
@@ -141,6 +149,7 @@ class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
       if (package != null) 'package': package,
       if (pluginClass != null) 'class': pluginClass,
       if (dartPluginClass != null) kDartPluginClass : dartPluginClass,
+      if (dartFileName != null) kDartFileName: dartFileName,
       if (ffiPlugin) kFfiPlugin: true,
       if (defaultPackage != null) kDefaultPackage : defaultPackage,
       // Mustache doesn't support complex types.
@@ -224,8 +233,8 @@ class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
 /// - an implementation consisting of:
 ///   - the [pluginClass] (with optional [classPrefix]) that will be the entry
 ///     point to the plugin's native code, and/or
-///   - the [dartPluginClass] that will be the entry point for the plugin's
-///     Dart code
+///   - the [dartPluginClass] with optional [dartFileName] that will be
+///     the entry point for the plugin's Dart code
 /// is required.
 class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlugin {
   const IOSPlugin({
@@ -233,6 +242,7 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
     required this.classPrefix,
     this.pluginClass,
     this.dartPluginClass,
+    this.dartFileName,
     bool? ffiPlugin,
     this.defaultPackage,
     bool? sharedDarwinSource,
@@ -244,6 +254,7 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
         classPrefix = '',
         pluginClass = yaml[kPluginClass] as String?,
         dartPluginClass = yaml[kDartPluginClass] as String?,
+        dartFileName = yaml[kDartFileName] as String?,
         ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
         defaultPackage = yaml[kDefaultPackage] as String?,
         sharedDarwinSource = yaml[kSharedDarwinSource] as bool? ?? false;
@@ -265,6 +276,7 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
   final String classPrefix;
   final String? pluginClass;
   final String? dartPluginClass;
+  final String? dartFileName;
   final bool ffiPlugin;
   final String? defaultPackage;
 
@@ -289,6 +301,7 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
       'prefix': classPrefix,
       if (pluginClass != null) 'class': pluginClass,
       if (dartPluginClass != null) kDartPluginClass : dartPluginClass,
+      if (dartFileName != null) kDartFileName : dartFileName,
       if (ffiPlugin) kFfiPlugin: true,
       if (sharedDarwinSource) kSharedDarwinSource: true,
       if (defaultPackage != null) kDefaultPackage : defaultPackage,
@@ -301,11 +314,14 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
 /// The [name] of the plugin is required. Either [dartPluginClass] or
 /// [pluginClass] or [ffiPlugin] are required.
 /// [pluginClass] will be the entry point to the plugin's native code.
+/// [dartFileName] is not required and will be used only if [dartPluginClass]
+/// provided.
 class MacOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlugin {
   const MacOSPlugin({
     required this.name,
     this.pluginClass,
     this.dartPluginClass,
+    this.dartFileName,
     bool? ffiPlugin,
     this.defaultPackage,
     bool? sharedDarwinSource,
@@ -317,6 +333,7 @@ class MacOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPl
         // Treat 'none' as not present. See https://github.com/flutter/flutter/issues/57497.
         pluginClass = yaml[kPluginClass] == 'none' ? null : yaml[kPluginClass] as String?,
         dartPluginClass = yaml[kDartPluginClass] as String?,
+        dartFileName = yaml[kDartFileName] as String?,
         ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
         defaultPackage = yaml[kDefaultPackage] as String?,
         sharedDarwinSource = yaml[kSharedDarwinSource] as bool? ?? false;
@@ -334,6 +351,7 @@ class MacOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPl
   final String name;
   final String? pluginClass;
   final String? dartPluginClass;
+  final String? dartFileName;
   final bool ffiPlugin;
   final String? defaultPackage;
 
@@ -357,6 +375,7 @@ class MacOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPl
       'name': name,
       if (pluginClass != null) 'class': pluginClass,
       if (dartPluginClass != null) kDartPluginClass: dartPluginClass,
+      if (dartFileName != null) kDartFileName: dartFileName,
       if (ffiPlugin) kFfiPlugin: true,
       if (sharedDarwinSource) kSharedDarwinSource: true,
       if (defaultPackage != null) kDefaultPackage: defaultPackage,
@@ -368,12 +387,15 @@ class MacOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPl
 ///
 /// The [name] of the plugin is required. Either [dartPluginClass] or [pluginClass] are required.
 /// [pluginClass] will be the entry point to the plugin's native code.
+/// [dartFileName] is not required and will be used only if [dartPluginClass]
+/// provided.
 class WindowsPlugin extends PluginPlatform
     implements NativeOrDartPlugin, VariantPlatformPlugin {
   const WindowsPlugin({
     required this.name,
     this.pluginClass,
     this.dartPluginClass,
+    this.dartFileName,
     bool? ffiPlugin,
     this.defaultPackage,
     this.variants = const <PluginPlatformVariant>{},
@@ -409,6 +431,7 @@ class WindowsPlugin extends PluginPlatform
       name: name,
       pluginClass: pluginClass,
       dartPluginClass: yaml[kDartPluginClass] as String?,
+      dartFileName: yaml[kDartFileName] as String?,
       ffiPlugin: yaml[kFfiPlugin] as bool?,
       defaultPackage: yaml[kDefaultPackage] as String?,
       variants: variants,
@@ -427,6 +450,7 @@ class WindowsPlugin extends PluginPlatform
   final String name;
   final String? pluginClass;
   final String? dartPluginClass;
+  final String? dartFileName;
   final bool ffiPlugin;
   final String? defaultPackage;
   final Set<PluginPlatformVariant> variants;
@@ -450,6 +474,7 @@ class WindowsPlugin extends PluginPlatform
       if (pluginClass != null) 'class': pluginClass,
       if (pluginClass != null) 'filename': _filenameForCppClass(pluginClass!),
       if (dartPluginClass != null) kDartPluginClass: dartPluginClass,
+      if (dartFileName != null) kDartFileName: dartFileName,
       if (ffiPlugin) kFfiPlugin: true,
       if (defaultPackage != null) kDefaultPackage: defaultPackage,
     };
@@ -460,11 +485,14 @@ class WindowsPlugin extends PluginPlatform
 ///
 /// The [name] of the plugin is required. Either [dartPluginClass] or [pluginClass] are required.
 /// [pluginClass] will be the entry point to the plugin's native code.
+/// [dartFileName] is not required and will be used only if [dartPluginClass]
+/// provided.
 class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
   const LinuxPlugin({
     required this.name,
     this.pluginClass,
     this.dartPluginClass,
+    this.dartFileName,
     bool? ffiPlugin,
     this.defaultPackage,
   })  : ffiPlugin = ffiPlugin ?? false,
@@ -475,6 +503,7 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
         // Treat 'none' as not present. See https://github.com/flutter/flutter/issues/57497.
         pluginClass = yaml[kPluginClass] == 'none' ? null : yaml[kPluginClass] as String?,
         dartPluginClass = yaml[kDartPluginClass] as String?,
+        dartFileName = yaml[kDartFileName] as String?,
         ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
         defaultPackage = yaml[kDefaultPackage] as String?;
 
@@ -490,6 +519,7 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
   final String name;
   final String? pluginClass;
   final String? dartPluginClass;
+  final String? dartFileName;
   final bool ffiPlugin;
   final String? defaultPackage;
 
@@ -509,6 +539,7 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
       if (pluginClass != null) 'class': pluginClass,
       if (pluginClass != null) 'filename': _filenameForCppClass(pluginClass!),
       if (dartPluginClass != null) kDartPluginClass: dartPluginClass,
+      if (dartFileName != null) kDartFileName: dartFileName,
       if (ffiPlugin) kFfiPlugin: true,
       if (defaultPackage != null) kDefaultPackage: defaultPackage,
     };

--- a/packages/flutter_tools/lib/src/platform_plugins.dart
+++ b/packages/flutter_tools/lib/src/platform_plugins.dart
@@ -88,15 +88,33 @@ class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
   })  : _fileSystem = fileSystem,
         ffiPlugin = ffiPlugin ?? false;
 
-  AndroidPlugin.fromYaml(this.name, YamlMap yaml, this.pluginPath, FileSystem fileSystem)
-      : assert(validate(yaml)),
-        package = yaml['package'] as String?,
-        pluginClass = yaml[kPluginClass] as String?,
-        dartPluginClass = yaml[kDartPluginClass] as String?,
-        dartFileName = yaml[kDartFileName] as String?,
-        ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
-        defaultPackage = yaml[kDefaultPackage] as String?,
-        _fileSystem = fileSystem;
+  factory AndroidPlugin.fromYaml(
+    String name,
+    YamlMap yaml,
+    String pluginPath,
+    FileSystem fileSystem,
+  ) {
+    assert(validate(yaml));
+
+    final String? dartPluginClass = yaml[kDartPluginClass] as String?;
+    final String? dartFileName = yaml[kDartFileName] as String?;
+
+    if (dartPluginClass == null && dartFileName != null) {
+      throw throwToolExit('"dartFileName" cannot be specified without "dartPluginClass" in Android platform of plugin "$name"');
+    }
+
+    return AndroidPlugin(
+      name: name,
+      package: yaml['package'] as String?,
+      pluginClass: yaml[kPluginClass] as String?,
+      dartPluginClass: dartPluginClass,
+      dartFileName: dartFileName,
+      ffiPlugin: yaml[kFfiPlugin] as bool? ?? false,
+      defaultPackage: yaml[kDefaultPackage] as String?,
+      pluginPath: pluginPath,
+      fileSystem: fileSystem,
+    );
+  }
 
   final FileSystem _fileSystem;
 
@@ -249,15 +267,27 @@ class IOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPlug
   }) : ffiPlugin = ffiPlugin ?? false,
        sharedDarwinSource = sharedDarwinSource ?? false;
 
-  IOSPlugin.fromYaml(this.name, YamlMap yaml)
-      : assert(validate(yaml)), // TODO(zanderso): https://github.com/flutter/flutter/issues/67241
-        classPrefix = '',
-        pluginClass = yaml[kPluginClass] as String?,
-        dartPluginClass = yaml[kDartPluginClass] as String?,
-        dartFileName = yaml[kDartFileName] as String?,
-        ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
-        defaultPackage = yaml[kDefaultPackage] as String?,
-        sharedDarwinSource = yaml[kSharedDarwinSource] as bool? ?? false;
+  factory IOSPlugin.fromYaml(String name, YamlMap yaml) {
+    assert(validate(yaml)); // TODO(zanderso): https://github.com/flutter/flutter/issues/67241
+
+    final String? dartPluginClass = yaml[kDartPluginClass] as String?;
+    final String? dartFileName = yaml[kDartFileName] as String?;
+
+    if (dartPluginClass == null && dartFileName != null) {
+      throwToolExit('"dartFileName" cannot be specified without "dartPluginClass" in iOS platform of plugin "$name"');
+    }
+
+    return IOSPlugin(
+      name: name,
+      classPrefix: '',
+      pluginClass: yaml[kPluginClass] as String?,
+      dartPluginClass: dartPluginClass,
+      dartFileName: dartFileName,
+      ffiPlugin: yaml[kFfiPlugin] as bool? ?? false,
+      defaultPackage: yaml[kDefaultPackage] as String?,
+      sharedDarwinSource: yaml[kSharedDarwinSource] as bool? ?? false,
+    );
+  }
 
   static bool validate(YamlMap yaml) {
     return yaml[kPluginClass] is String ||
@@ -328,15 +358,29 @@ class MacOSPlugin extends PluginPlatform implements NativeOrDartPlugin, DarwinPl
   }) : ffiPlugin = ffiPlugin ?? false,
        sharedDarwinSource = sharedDarwinSource ?? false;
 
-  MacOSPlugin.fromYaml(this.name, YamlMap yaml)
-      : assert(validate(yaml)),
-        // Treat 'none' as not present. See https://github.com/flutter/flutter/issues/57497.
-        pluginClass = yaml[kPluginClass] == 'none' ? null : yaml[kPluginClass] as String?,
-        dartPluginClass = yaml[kDartPluginClass] as String?,
-        dartFileName = yaml[kDartFileName] as String?,
-        ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
-        defaultPackage = yaml[kDefaultPackage] as String?,
-        sharedDarwinSource = yaml[kSharedDarwinSource] as bool? ?? false;
+  factory MacOSPlugin.fromYaml(String name, YamlMap yaml) {
+    assert(validate(yaml));
+
+    final String? dartPluginClass = yaml[kDartPluginClass] as String?;
+    final String? dartFileName = yaml[kDartFileName] as String?;
+
+    if (dartPluginClass == null && dartFileName != null) {
+      throwToolExit('"dartFileName" cannot be specified without "dartPluginClass" in macOS platform of plugin "$name"');
+    }
+
+    // Treat 'none' as not present. See https://github.com/flutter/flutter/issues/57497.
+    final String? pluginClass = yaml[kPluginClass] == 'none' ? null : yaml[kPluginClass] as String?;
+
+    return MacOSPlugin(
+      name: name,
+      pluginClass: pluginClass,
+      dartPluginClass: dartPluginClass,
+      dartFileName: dartFileName,
+      ffiPlugin: yaml[kFfiPlugin] as bool?,
+      defaultPackage: yaml[kDefaultPackage] as String?,
+      sharedDarwinSource: yaml[kSharedDarwinSource] as bool?,
+    );
+  }
 
   static bool validate(YamlMap yaml) {
     return yaml[kPluginClass] is String ||
@@ -427,11 +471,18 @@ class WindowsPlugin extends PluginPlatform
         // future non-breaking.
       }
     }
+
+    final String? dartPluginClass = yaml[kDartPluginClass] as String?;
+    final String? dartFileName = yaml[kDartFileName] as String?;
+
+    if (dartPluginClass == null && dartFileName != null) {
+      throwToolExit('"dartFileName" cannot be specified without "dartPluginClass" in Windows platform of plugin "$name"');
+    }
     return WindowsPlugin(
       name: name,
       pluginClass: pluginClass,
-      dartPluginClass: yaml[kDartPluginClass] as String?,
-      dartFileName: yaml[kDartFileName] as String?,
+      dartPluginClass: dartPluginClass,
+      dartFileName: dartFileName,
       ffiPlugin: yaml[kFfiPlugin] as bool?,
       defaultPackage: yaml[kDefaultPackage] as String?,
       variants: variants,
@@ -498,14 +549,26 @@ class LinuxPlugin extends PluginPlatform implements NativeOrDartPlugin {
   })  : ffiPlugin = ffiPlugin ?? false,
         assert(pluginClass != null || dartPluginClass != null || (ffiPlugin ?? false) || defaultPackage != null);
 
-  LinuxPlugin.fromYaml(this.name, YamlMap yaml)
-      : assert(validate(yaml)),
-        // Treat 'none' as not present. See https://github.com/flutter/flutter/issues/57497.
-        pluginClass = yaml[kPluginClass] == 'none' ? null : yaml[kPluginClass] as String?,
-        dartPluginClass = yaml[kDartPluginClass] as String?,
-        dartFileName = yaml[kDartFileName] as String?,
-        ffiPlugin = yaml[kFfiPlugin] as bool? ?? false,
-        defaultPackage = yaml[kDefaultPackage] as String?;
+  factory LinuxPlugin.fromYaml(String name, YamlMap yaml) {
+    assert(validate(yaml));
+
+    final String? dartPluginClass = yaml[kDartPluginClass] as String?;
+    final String? dartFileName = yaml[kDartFileName] as String?;
+
+    if (dartPluginClass == null && dartFileName != null) {
+      throwToolExit('"dartFileName" cannot be specified without "dartPluginClass" in Linux platform of plugin "$name"');
+    }
+
+    return LinuxPlugin(
+      name: name,
+      // Treat 'none' as not present. See https://github.com/flutter/flutter/issues/57497.
+      pluginClass: yaml[kPluginClass] == 'none' ? null : yaml[kPluginClass] as String?,
+      dartPluginClass: dartPluginClass,
+      dartFileName: dartFileName,
+      ffiPlugin: yaml[kFfiPlugin] as bool? ?? false,
+      defaultPackage: yaml[kDefaultPackage] as String?,
+    );
+  }
 
   static bool validate(YamlMap yaml) {
     return yaml[kPluginClass] is String ||

--- a/packages/flutter_tools/lib/src/plugins.dart
+++ b/packages/flutter_tools/lib/src/plugins.dart
@@ -2,7 +2,6 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'package:meta/meta.dart';
 import 'package:pub_semver/pub_semver.dart';
 import 'package:yaml/yaml.dart';
 
@@ -167,8 +166,8 @@ class Plugin {
         }
         final DartPluginClassAndFilePair? dartPair = _getPluginDartClassForPlatform(
           platformsYaml,
-          platform,
-          name,
+          platformKey: platform,
+          pluginName: name,
         );
         if (dartPair != null) {
           dartPluginClasses[platform] = dartPair;
@@ -355,10 +354,10 @@ class Plugin {
   }
 
   static DartPluginClassAndFilePair? _getPluginDartClassForPlatform(
-    YamlMap platformsYaml,
-    String platformKey,
-    String name,
-  ) {
+    YamlMap platformsYaml,{
+    required String platformKey,
+    required String pluginName,
+  }) {
     if (!_supportsPlatform(platformsYaml, platformKey)) {
       return null;
     }
@@ -366,8 +365,8 @@ class Plugin {
       final String dartClass = (platformsYaml[platformKey] as YamlMap)[kDartPluginClass] as String;
       final String dartFileName = (platformsYaml[platformKey] as YamlMap)[kDartFileName]
         as String?
-        ?? '$name.dart';
-      return DartPluginClassAndFilePair(dartClass, dartFileName);
+        ?? '$pluginName.dart';
+      return (dartClass: dartClass, dartFileName: dartFileName);
     }
     return null;
   }
@@ -481,35 +480,13 @@ class PluginInterfaceResolution {
   }
 }
 
-/// A class representing pair of dartPluginClass and dartFileName used as metadata
+/// A record representing pair of dartPluginClass and dartFileName used as metadata
 /// in [PluginInterfaceResolution].
 ///
 /// The [dartClass] and [dartFileName] fields are guaranteed to be non-null:
-/// - instance of this class should be created only if dartClassName
-/// exists in plugin configuration.
+/// - record should be created only if dartClassName exists in plugin configuration.
 /// - dartFileName either taken from configuration, or, if absent, should be
 /// constructed from plugin name.
 /// See also:
-/// - [PluginInterfaceResolution], which uses this class to create Map with metadata.
-@immutable
-class DartPluginClassAndFilePair {
-  const DartPluginClassAndFilePair(this.dartClass, this.dartFileName);
-
-  final String dartClass;
-  final String dartFileName;
-
-  @override
-  bool operator ==(Object other) {
-    if (runtimeType != other.runtimeType) {
-      return false;
-    }
-    return other is DartPluginClassAndFilePair
-        && dartClass == other.dartClass
-        && dartFileName == other.dartFileName;
-  }
-  @override
-  int get hashCode => Object.hash(dartClass, dartFileName);
-
-  @override
-  String toString() => 'DartPluginClassAndFilePair(dartClass: $dartClass, dartFileName: $dartFileName)';
-}
+/// - [PluginInterfaceResolution], which uses this record to create Map with metadata.
+typedef DartPluginClassAndFilePair = ({String dartClass, String dartFileName});

--- a/packages/flutter_tools/lib/src/plugins.dart
+++ b/packages/flutter_tools/lib/src/plugins.dart
@@ -494,6 +494,7 @@ class PluginInterfaceResolution {
 @immutable
 class DartPluginClassAndFilePair {
   const DartPluginClassAndFilePair(this.dartClass, this.dartFileName);
+
   final String dartClass;
   final String dartFileName;
 

--- a/packages/flutter_tools/lib/src/plugins.dart
+++ b/packages/flutter_tools/lib/src/plugins.dart
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:meta/meta.dart';
 import 'package:pub_semver/pub_semver.dart';
 import 'package:yaml/yaml.dart';
 
@@ -48,6 +49,12 @@ class Plugin {
   ///          linux:
   ///            # A plugin implemented purely in Dart code.
   ///            dartPluginClass: SamplePlugin
+  ///            # Optional field to determine file containing dartPluginClass.
+  ///            # This file will be used in imports in generated files, e.g.:
+  ///            #   import 'package:{{pluginName}}/{{dartFileName}}'
+  ///            # instead of default:
+  ///            #    import 'package:{{pluginName}}/{{pluginName}}.dart'
+  ///            dartFileName: src/sample_plugin.dart
   ///          macos:
   ///            # A plugin implemented with `dart:ffi`.
   ///            ffiPlugin: true
@@ -152,15 +159,19 @@ class Plugin {
       WindowsPlugin.kConfigKey,
     ];
     final Map<String, String> defaultPackages = <String, String>{};
-    final Map<String, String> dartPluginClasses = <String, String>{};
+    final Map<String, DartPluginClassAndFilePair> dartPluginClasses = <String, DartPluginClassAndFilePair>{};
     for (final String platform in sharedHandlingPlatforms) {
         final String? defaultPackage = _getDefaultPackageForPlatform(platformsYaml, platform);
         if (defaultPackage != null) {
           defaultPackages[platform] = defaultPackage;
         }
-        final String? dartClass = _getPluginDartClassForPlatform(platformsYaml, platform);
-        if (dartClass != null) {
-          dartPluginClasses[platform] = dartClass;
+        final DartPluginClassAndFilePair? dartPair = _getPluginDartClassForPlatform(
+          platformsYaml,
+          platform,
+          name,
+        );
+        if (dartPair != null) {
+          dartPluginClasses[platform] = dartPair;
         }
     }
 
@@ -213,7 +224,7 @@ class Plugin {
       path: path,
       platforms: platforms,
       defaultPackagePlatforms: <String, String>{},
-      pluginDartClassPlatforms: <String, String>{},
+      pluginDartClassPlatforms: <String, DartPluginClassAndFilePair>{},
       flutterConstraint: flutterConstraint,
       dependencies: dependencies,
       isDirectDependency: isDirectDependency,
@@ -343,12 +354,20 @@ class Plugin {
     return null;
   }
 
-  static String? _getPluginDartClassForPlatform(YamlMap platformsYaml, String platformKey) {
+  static DartPluginClassAndFilePair? _getPluginDartClassForPlatform(
+    YamlMap platformsYaml,
+    String platformKey,
+    String name,
+  ) {
     if (!_supportsPlatform(platformsYaml, platformKey)) {
       return null;
     }
     if ((platformsYaml[platformKey] as YamlMap).containsKey(kDartPluginClass)) {
-      return (platformsYaml[platformKey] as YamlMap)[kDartPluginClass] as String;
+      final String dartClass = (platformsYaml[platformKey] as YamlMap)[kDartPluginClass] as String;
+      final String dartFileName = (platformsYaml[platformKey] as YamlMap)[kDartFileName]
+        as String?
+        ?? '$name.dart';
+      return DartPluginClassAndFilePair(dartClass, dartFileName);
     }
     return null;
   }
@@ -383,7 +402,7 @@ class Plugin {
   final Map<String, String> defaultPackagePlatforms;
 
   /// This is a mapping from platform config key to the Dart plugin class for the given platform.
-  final Map<String, String> pluginDartClassPlatforms;
+  final Map<String, DartPluginClassAndFilePair> pluginDartClassPlatforms;
 
   /// Whether this plugin is a direct dependency of the app.
   /// If [false], the plugin is a dependency of another plugin.
@@ -451,7 +470,8 @@ class PluginInterfaceResolution {
     return <String, String> {
       'pluginName': plugin.name,
       'platform': platform,
-      'dartClass': plugin.pluginDartClassPlatforms[platform] ?? '',
+      'dartClass': plugin.pluginDartClassPlatforms[platform]?.dartClass ?? '',
+      'dartFileName': plugin.pluginDartClassPlatforms[platform]?.dartFileName ?? '',
     };
   }
 
@@ -459,4 +479,36 @@ class PluginInterfaceResolution {
   String toString() {
     return '<PluginInterfaceResolution ${plugin.name} for $platform>';
   }
+}
+
+/// A class representing pair of dartPluginClass and dartFileName used as metadata
+/// in [PluginInterfaceResolution].
+///
+/// The [dartClass] and [dartFileName] fields are guaranteed to be non-null:
+/// - instance of this class should be created only if dartClassName
+/// exists in plugin configuration.
+/// - dartFileName either taken from configuration, or, if absent, should be
+/// constructed from plugin name.
+/// See also:
+/// - [PluginInterfaceResolution], which uses this class to create Map with metadata.
+@immutable
+class DartPluginClassAndFilePair {
+  const DartPluginClassAndFilePair(this.dartClass, this.dartFileName);
+  final String dartClass;
+  final String dartFileName;
+
+  @override
+  bool operator ==(Object other) {
+    if (runtimeType != other.runtimeType) {
+      return false;
+    }
+    return other is DartPluginClassAndFilePair
+        && dartClass == other.dartClass
+        && dartFileName == other.dartFileName;
+  }
+  @override
+  int get hashCode => Object.hash(dartClass, dartFileName);
+
+  @override
+  String toString() => 'DartPluginClassAndFilePair(dartClass: $dartClass, dartFileName: $dartFileName)';
 }

--- a/packages/flutter_tools/test/general.shard/dart_plugin_test.dart
+++ b/packages/flutter_tools/test/general.shard/dart_plugin_test.dart
@@ -92,6 +92,7 @@ void main() {
             'pluginName': 'url_launcher_linux',
             'dartClass': 'UrlLauncherPluginLinux',
             'platform': 'linux',
+            'dartFileName': 'url_launcher_linux.dart'
           })
         );
         expect(resolutions[1].toMap(), equals(
@@ -99,6 +100,7 @@ void main() {
             'pluginName': 'url_launcher_macos',
             'dartClass': 'UrlLauncherPluginMacOS',
             'platform': 'macos',
+            'dartFileName': 'url_launcher_macos.dart',
           })
         );
       });
@@ -157,6 +159,7 @@ void main() {
               'pluginName': 'url_launcher_macos',
               'dartClass': 'UrlLauncherPluginMacOS',
               'platform': 'macos',
+              'dartFileName': 'url_launcher_macos.dart',
             }));
       });
 
@@ -208,6 +211,7 @@ void main() {
             'pluginName': 'url_launcher_macos',
             'dartClass': 'UrlLauncherPluginMacOS',
             'platform': 'macos',
+            'dartFileName': 'url_launcher_macos.dart',
           })
         );
         expect(resolutions[1].toMap(), equals(
@@ -215,6 +219,7 @@ void main() {
             'pluginName': 'transitive_dependency_plugin',
             'dartClass': 'UrlLauncherPluginWindows',
             'platform': 'windows',
+            'dartFileName': 'transitive_dependency_plugin.dart'
           })
         );
       });
@@ -251,6 +256,7 @@ void main() {
             'pluginName': 'url_launcher',
             'dartClass': 'UrlLauncherAndroid',
             'platform': 'android',
+            'dartFileName': 'url_launcher.dart',
           })
         );
         expect(resolutions[1].toMap(), equals(
@@ -258,6 +264,7 @@ void main() {
             'pluginName': 'url_launcher',
             'dartClass': 'UrlLauncherIos',
             'platform': 'ios',
+            'dartFileName': 'url_launcher.dart',
           })
         );
       });
@@ -368,16 +375,19 @@ void main() {
               'pluginName': 'url_launcher',
               'dartClass': 'UrlLauncherLinux',
               'platform': 'linux',
+              'dartFileName': 'url_launcher.dart',
             },
             <String, String>{
               'pluginName': 'url_launcher',
               'dartClass': 'UrlLauncherMacOS',
               'platform': 'macos',
+              'dartFileName': 'url_launcher.dart',
             },
             <String, String>{
               'pluginName': 'url_launcher',
               'dartClass': 'UrlLauncherWindows',
               'platform': 'windows',
+              'dartFileName': 'url_launcher.dart',
             },
           ])
         );
@@ -463,6 +473,7 @@ void main() {
             'pluginName': 'url_launcher_linux',
             'dartClass': 'UrlLauncherPluginLinux',
             'platform': 'linux',
+            'dartFileName': 'url_launcher_linux.dart',
           })
         );
       });
@@ -512,6 +523,7 @@ void main() {
             'pluginName': 'url_launcher_linux',
             'dartClass': 'UrlLauncherPluginLinux',
             'platform': 'linux',
+            'dartFileName': 'url_launcher_linux.dart'
           })
         );
       });
@@ -580,6 +592,7 @@ void main() {
             'pluginName': 'user_selected_url_launcher_implementation',
             'dartClass': 'UrlLauncherPluginLinux',
             'platform': 'linux',
+            'dartFileName': 'user_selected_url_launcher_implementation.dart',
           })
         );
       });
@@ -635,6 +648,7 @@ void main() {
               'pluginName': 'user_selected_url_launcher_implementation',
               'dartClass': 'UrlLauncherAndroid',
               'platform': 'android',
+              'dartFileName': 'user_selected_url_launcher_implementation.dart',
             })
         );
         expect(resolutions[1].toMap(), equals(
@@ -642,6 +656,7 @@ void main() {
               'pluginName': 'url_launcher',
               'dartClass': 'UrlLauncherIos',
               'platform': 'ios',
+              'dartFileName': 'url_launcher.dart'
             })
         );
       });
@@ -914,6 +929,7 @@ void main() {
               'pluginName': 'url_launcher_linux',
               'dartClass': 'UrlLauncherLinux',
               'platform': 'linux',
+              'dartFileName': 'url_launcher_linux.dart'
             })
         );
       });

--- a/packages/flutter_tools/test/general.shard/dart_plugin_test.dart
+++ b/packages/flutter_tools/test/general.shard/dart_plugin_test.dart
@@ -92,7 +92,7 @@ void main() {
             'pluginName': 'url_launcher_linux',
             'dartClass': 'UrlLauncherPluginLinux',
             'platform': 'linux',
-            'dartFileName': 'url_launcher_linux.dart'
+            'dartFileName': 'url_launcher_linux.dart',
           })
         );
         expect(resolutions[1].toMap(), equals(
@@ -219,7 +219,7 @@ void main() {
             'pluginName': 'transitive_dependency_plugin',
             'dartClass': 'UrlLauncherPluginWindows',
             'platform': 'windows',
-            'dartFileName': 'transitive_dependency_plugin.dart'
+            'dartFileName': 'transitive_dependency_plugin.dart',
           })
         );
       });
@@ -523,7 +523,7 @@ void main() {
             'pluginName': 'url_launcher_linux',
             'dartClass': 'UrlLauncherPluginLinux',
             'platform': 'linux',
-            'dartFileName': 'url_launcher_linux.dart'
+            'dartFileName': 'url_launcher_linux.dart',
           })
         );
       });
@@ -656,7 +656,7 @@ void main() {
               'pluginName': 'url_launcher',
               'dartClass': 'UrlLauncherIos',
               'platform': 'ios',
-              'dartFileName': 'url_launcher.dart'
+              'dartFileName': 'url_launcher.dart',
             })
         );
       });
@@ -929,7 +929,7 @@ void main() {
               'pluginName': 'url_launcher_linux',
               'dartClass': 'UrlLauncherLinux',
               'platform': 'linux',
-              'dartFileName': 'url_launcher_linux.dart'
+              'dartFileName': 'url_launcher_linux.dart',
             })
         );
       });

--- a/packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
+++ b/packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
@@ -214,46 +214,134 @@ void main() {
     );
 
     expect(plugin.pluginDartClassPlatforms, <String, DartPluginClassAndFilePair>{
-      'android': const DartPluginClassAndFilePair('AndroidClass', 'src/android_class.dart'),
-      'ios': const DartPluginClassAndFilePair('IosClass', 'src/ios_class.dart'),
-      'linux':  const DartPluginClassAndFilePair('LinuxClass', 'src/linux_class.dart'),
-      'macos':  const DartPluginClassAndFilePair('MacOSClass', 'src/macos_class.dart'),
-      'windows':  const DartPluginClassAndFilePair('WindowsClass', 'src/windows_class.dart'),
+      'android': const (dartClass: 'AndroidClass', dartFileName: 'src/android_class.dart'),
+      'ios': const (dartClass: 'IosClass', dartFileName: 'src/ios_class.dart'),
+      'linux':  const (dartClass: 'LinuxClass', dartFileName: 'src/linux_class.dart'),
+      'macos':  const (dartClass: 'MacOSClass', dartFileName: 'src/macos_class.dart'),
+      'windows':  const (dartClass: 'WindowsClass', dartFileName: 'src/windows_class.dart'),
     });
   });
 
-  testWithoutContext('dartFileName without dartPluginClass has no effect', () {
+  testWithoutContext('dartFileName without dartPluginClass throws (Android)', () {
     final FileSystem fileSystem = MemoryFileSystem.test();
     const String pluginYamlRaw =
       'platforms:\n'
       ' android:\n'
       '  package: com.example\n'
       '  pluginClass: AndroidClass\n'
-      '  dartFileName: src/android_class.dart\n'
+      '  dartFileName: src/android_class.dart\n';
+
+    final YamlMap pluginYaml = loadYaml(pluginYamlRaw) as YamlMap;
+    expect(
+     () => Plugin.fromYaml(
+        _kTestPluginName,
+        _kTestPluginPath,
+        pluginYaml,
+        null,
+        const <String>[],
+        fileSystem: fileSystem,
+      ),
+      throwsToolExit(
+        message: '"dartFileName" cannot be specified without "dartPluginClass" in Android platform of plugin "$_kTestPluginName"',
+      ),
+    );
+  });
+
+  testWithoutContext('dartFileName without dartPluginClass throws (iOS)', () {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    const String pluginYamlRaw =
+      'platforms:\n'
       ' ios:\n'
       '  pluginClass: IosClass\n'
-      '  dartFileName: src/ios_class.dart\n'
+      '  dartFileName: src/ios_class.dart\n';
+
+    final YamlMap pluginYaml = loadYaml(pluginYamlRaw) as YamlMap;
+    expect(
+     () => Plugin.fromYaml(
+        _kTestPluginName,
+        _kTestPluginPath,
+        pluginYaml,
+        null,
+        const <String>[],
+        fileSystem: fileSystem,
+      ),
+      throwsToolExit(
+        message: '"dartFileName" cannot be specified without "dartPluginClass" in iOS platform of plugin "$_kTestPluginName"',
+      ),
+    );
+  });
+
+  testWithoutContext('dartFileName without dartPluginClass throws (Linux)', () {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    const String pluginYamlRaw =
+      'platforms:\n'
       ' linux:\n'
       '  pluginClass: LinuxClass\n'
-      '  dartFileName: src/linux_class.dart\n'
+      '  dartFileName: src/linux_class.dart\n';
+
+    final YamlMap pluginYaml = loadYaml(pluginYamlRaw) as YamlMap;
+    expect(
+     () => Plugin.fromYaml(
+        _kTestPluginName,
+        _kTestPluginPath,
+        pluginYaml,
+        null,
+        const <String>[],
+        fileSystem: fileSystem,
+      ),
+      throwsToolExit(
+        message: '"dartFileName" cannot be specified without "dartPluginClass" in Linux platform of plugin "$_kTestPluginName"',
+      ),
+    );
+  });
+
+  testWithoutContext('dartFileName without dartPluginClass throws (MacOS)', () {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    const String pluginYamlRaw =
+      'platforms:\n'
       ' macos:\n'
       '  pluginClass: MacOSClass\n'
-      '  dartFileName: src/macos_class.dart\n'
+      '  dartFileName: src/macos_class.dart\n';
+
+    final YamlMap pluginYaml = loadYaml(pluginYamlRaw) as YamlMap;
+    expect(
+     () => Plugin.fromYaml(
+        _kTestPluginName,
+        _kTestPluginPath,
+        pluginYaml,
+        null,
+        const <String>[],
+        fileSystem: fileSystem,
+      ),
+      throwsToolExit(
+        message: '"dartFileName" cannot be specified without "dartPluginClass" in macOS platform of plugin "$_kTestPluginName"',
+      ),
+    );
+  });
+
+
+  testWithoutContext('dartFileName without dartPluginClass throws (Windows)', () {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    const String pluginYamlRaw =
+      'platforms:\n'
       ' windows:\n'
       '  pluginClass: WindowsClass\n'
       '  dartFileName: src/windows_class.dart\n';
 
     final YamlMap pluginYaml = loadYaml(pluginYamlRaw) as YamlMap;
-    final Plugin plugin = Plugin.fromYaml(
-      _kTestPluginName,
-      _kTestPluginPath,
-      pluginYaml,
-      null,
-      const <String>[],
-      fileSystem: fileSystem,
+    expect(
+     () => Plugin.fromYaml(
+        _kTestPluginName,
+        _kTestPluginPath,
+        pluginYaml,
+        null,
+        const <String>[],
+        fileSystem: fileSystem,
+      ),
+      throwsToolExit(
+        message: '"dartFileName" cannot be specified without "dartPluginClass" in Windows platform of plugin "$_kTestPluginName"',
+      ),
     );
-
-    expect(plugin.pluginDartClassPlatforms, <String, DartPluginClassAndFilePair>{});
   });
 
   testWithoutContext('Plugin parsing of legacy format and multi-platform format together is not allowed '
@@ -339,9 +427,9 @@ void main() {
     );
 
     expect(plugin.pluginDartClassPlatforms, <String, DartPluginClassAndFilePair>{
-      'linux':  const DartPluginClassAndFilePair('LinuxClass', 'test_plugin_name.dart'),
-      'macos':  const DartPluginClassAndFilePair('MacOSClass', 'test_plugin_name.dart'),
-      'windows':  const DartPluginClassAndFilePair('WindowsClass', 'test_plugin_name.dart'),
+      'linux':  const (dartClass: 'LinuxClass', dartFileName: 'test_plugin_name.dart'),
+      'macos':  const (dartClass: 'MacOSClass', dartFileName: 'test_plugin_name.dart'),
+      'windows':  const (dartClass: 'WindowsClass', dartFileName: 'test_plugin_name.dart'),
     });
   });
 

--- a/packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
+++ b/packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
@@ -183,6 +183,79 @@ void main() {
     expect(windowsPlugin.dartPluginClass, 'WinSamplePlugin');
   });
 
+  testWithoutContext('Plugin parsing allows a dartFileName field with dartPluginClass', () {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    const String pluginYamlRaw =
+      'platforms:\n'
+      ' android:\n'
+      '  dartPluginClass: AndroidClass\n'
+      '  dartFileName: src/android_class.dart\n'
+      ' ios:\n'
+      '  dartPluginClass: IosClass\n'
+      '  dartFileName: src/ios_class.dart\n'
+      ' linux:\n'
+      '  dartPluginClass: LinuxClass\n'
+      '  dartFileName: src/linux_class.dart\n'
+      ' macos:\n'
+      '  dartPluginClass: MacOSClass\n'
+      '  dartFileName: src/macos_class.dart\n'
+      ' windows:\n'
+      '  dartPluginClass: WindowsClass\n'
+      '  dartFileName: src/windows_class.dart\n';
+
+    final YamlMap pluginYaml = loadYaml(pluginYamlRaw) as YamlMap;
+    final Plugin plugin = Plugin.fromYaml(
+      _kTestPluginName,
+      _kTestPluginPath,
+      pluginYaml,
+      null,
+      const <String>[],
+      fileSystem: fileSystem,
+    );
+
+    expect(plugin.pluginDartClassPlatforms, <String, DartPluginClassAndFilePair>{
+      'android': const DartPluginClassAndFilePair('AndroidClass', 'src/android_class.dart'),
+      'ios': const DartPluginClassAndFilePair('IosClass', 'src/ios_class.dart'),
+      'linux':  const DartPluginClassAndFilePair('LinuxClass', 'src/linux_class.dart'),
+      'macos':  const DartPluginClassAndFilePair('MacOSClass', 'src/macos_class.dart'),
+      'windows':  const DartPluginClassAndFilePair('WindowsClass', 'src/windows_class.dart'),
+    });
+  });
+
+  testWithoutContext('dartFileName without dartPluginClass has no effect', () {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    const String pluginYamlRaw =
+      'platforms:\n'
+      ' android:\n'
+      '  package: com.example\n'
+      '  pluginClass: AndroidClass\n'
+      '  dartFileName: src/android_class.dart\n'
+      ' ios:\n'
+      '  pluginClass: IosClass\n'
+      '  dartFileName: src/ios_class.dart\n'
+      ' linux:\n'
+      '  pluginClass: LinuxClass\n'
+      '  dartFileName: src/linux_class.dart\n'
+      ' macos:\n'
+      '  pluginClass: MacOSClass\n'
+      '  dartFileName: src/macos_class.dart\n'
+      ' windows:\n'
+      '  pluginClass: WindowsClass\n'
+      '  dartFileName: src/windows_class.dart\n';
+
+    final YamlMap pluginYaml = loadYaml(pluginYamlRaw) as YamlMap;
+    final Plugin plugin = Plugin.fromYaml(
+      _kTestPluginName,
+      _kTestPluginPath,
+      pluginYaml,
+      null,
+      const <String>[],
+      fileSystem: fileSystem,
+    );
+
+    expect(plugin.pluginDartClassPlatforms, <String, DartPluginClassAndFilePair>{});
+  });
+
   testWithoutContext('Plugin parsing of legacy format and multi-platform format together is not allowed '
     'and fatal error message contains plugin name', () {
     final FileSystem fileSystem = MemoryFileSystem.test();

--- a/packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
+++ b/packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
@@ -265,10 +265,10 @@ void main() {
       fileSystem: fileSystem,
     );
 
-    expect(plugin.pluginDartClassPlatforms, <String, String>{
-      'linux': 'LinuxClass',
-      'macos': 'MacOSClass',
-      'windows': 'WindowsClass',
+    expect(plugin.pluginDartClassPlatforms, <String, DartPluginClassAndFilePair>{
+      'linux':  const DartPluginClassAndFilePair('LinuxClass', 'test_plugin_name.dart'),
+      'macos':  const DartPluginClassAndFilePair('MacOSClass', 'test_plugin_name.dart'),
+      'windows':  const DartPluginClassAndFilePair('WindowsClass', 'test_plugin_name.dart'),
     });
   });
 

--- a/packages/flutter_tools/test/general.shard/plugins_test.dart
+++ b/packages/flutter_tools/test/general.shard/plugins_test.dart
@@ -1684,7 +1684,7 @@ The Flutter Preview device does not support the following plugins from your pubs
           name: 'test',
           path: '/path/to/test/',
           defaultPackagePlatforms: const <String, String>{},
-          pluginDartClassPlatforms: const <String, String>{},
+          pluginDartClassPlatforms: const <String, DartPluginClassAndFilePair>{},
           platforms: const <String, PluginPlatform>{
             IOSPlugin.kConfigKey: IOSPlugin(name: 'test', classPrefix: ''),
             MacOSPlugin.kConfigKey: MacOSPlugin(name: 'test'),
@@ -1709,7 +1709,7 @@ The Flutter Preview device does not support the following plugins from your pubs
           name: 'test',
           path: '/path/to/test/',
           defaultPackagePlatforms: const <String, String>{},
-          pluginDartClassPlatforms: const <String, String>{},
+          pluginDartClassPlatforms: const <String, DartPluginClassAndFilePair>{},
           platforms: const <String, PluginPlatform>{
             IOSPlugin.kConfigKey: IOSPlugin(name: 'test', classPrefix: '', sharedDarwinSource: true),
             MacOSPlugin.kConfigKey: MacOSPlugin(name: 'test', sharedDarwinSource: true),
@@ -1734,7 +1734,7 @@ The Flutter Preview device does not support the following plugins from your pubs
           name: 'test',
           path: '/path/to/test/',
           defaultPackagePlatforms: const <String, String>{},
-          pluginDartClassPlatforms: const <String, String>{},
+          pluginDartClassPlatforms: const <String, DartPluginClassAndFilePair>{},
           platforms: const <String, PluginPlatform>{
             WindowsPlugin.kConfigKey: WindowsPlugin(name: 'test', pluginClass: ''),
           },
@@ -1762,7 +1762,7 @@ The Flutter Preview device does not support the following plugins from your pubs
           name: 'test',
           path: '/path/to/test/',
           defaultPackagePlatforms: const <String, String>{},
-          pluginDartClassPlatforms: const <String, String>{},
+          pluginDartClassPlatforms: const <String, DartPluginClassAndFilePair>{},
           platforms: const <String, PluginPlatform>{
             IOSPlugin.kConfigKey: IOSPlugin(name: 'test', classPrefix: ''),
             MacOSPlugin.kConfigKey: MacOSPlugin(name: 'test'),
@@ -1787,7 +1787,7 @@ The Flutter Preview device does not support the following plugins from your pubs
           name: 'test',
           path: '/path/to/test/',
           defaultPackagePlatforms: const <String, String>{},
-          pluginDartClassPlatforms: const <String, String>{},
+          pluginDartClassPlatforms: const <String, DartPluginClassAndFilePair>{},
           platforms: const <String, PluginPlatform>{
             IOSPlugin.kConfigKey: IOSPlugin(name: 'test', classPrefix: '', sharedDarwinSource: true),
             MacOSPlugin.kConfigKey: MacOSPlugin(name: 'test', sharedDarwinSource: true),
@@ -1812,7 +1812,7 @@ The Flutter Preview device does not support the following plugins from your pubs
           name: 'test',
           path: '/path/to/test/',
           defaultPackagePlatforms: const <String, String>{},
-          pluginDartClassPlatforms: const <String, String>{},
+          pluginDartClassPlatforms: const <String, DartPluginClassAndFilePair>{},
           platforms: const <String, PluginPlatform>{
             WindowsPlugin.kConfigKey: WindowsPlugin(name: 'test', pluginClass: ''),
           },

--- a/packages/flutter_tools/test/general.shard/windows/plugins_test.dart
+++ b/packages/flutter_tools/test/general.shard/windows/plugins_test.dart
@@ -28,7 +28,7 @@ void main() {
         name: 'test',
         path: 'foo',
         defaultPackagePlatforms: const <String, String>{},
-        pluginDartClassPlatforms: const <String, String>{},
+        pluginDartClassPlatforms: const <String, DartPluginClassAndFilePair>{},
         platforms: const <String, PluginPlatform>{
           WindowsPlugin.kConfigKey: WindowsPlugin(
             name: 'test',

